semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/SymbolOps.scala

```diff
@@ -0,0 +1,244 @@
+        sym.infoType match {
+          case TypeRefType(_, moduleClass, _) => moduleClass
+          case _ => NoSymbol
+        }
+      case _ =>
+        NoSymbol
+    }
+    def isClass: Boolean = sym.isInstanceOf[ClassSymbol] && !sym.isModule
+    def isObject: Boolean = sym.isInstanceOf[ObjectSymbol]
+    def isType: Boolean = sym.isInstanceOf[TypeSymbol]
+    def isAlias: Boolean = sym.isInstanceOf[AliasSymbol]
+    def isMacro: Boolean = sym.isMethod && sym.hasFlag(0x00008000)
+    def isConstructor: Boolean = sym.isMethod && (sym.name == "<init>" || sym.name == "$init$")
+    def isPackageObject: Boolean = sym.name == "package"
+    def isTypeParam = sym.isType && sym.isParam
+    def isAnonymousClass = sym.name.contains("$anon")
+    def isAnonymousFunction = sym.name.contains("$anonfun")
+    def isSyntheticConstructor = sym match {
+      case sym: SymbolInfoSymbol =>
+        val owner = sym.symbolInfo.owner
+        sym.isConstructor && (owner.isModuleClass || owner.isTrait)
+      case _ =>
+        false
+    }
+    def isLocalChild: Boolean = sym.name == "<local child>"
+    def isExtensionMethod: Boolean = sym.name.contains("$extension")
+    def isSyntheticValueClassCompanion: Boolean = {
+      sym match {
+        case sym: SymbolInfoSymbol =>
+          if (sym.isModuleClass) {
+            sym.infoType match {
+              case ClassInfoType(_, List(TypeRefType(_, anyRef, _))) =>
+                sym.isSynthetic && sym.semanticdbDecls.syms.isEmpty
+              case _ =>
+                false
+            }
+          } else if (sym.isModule) {
+            sym.moduleClass.isSyntheticValueClassCompanion
+          } else {
+            false
+          }
+        case _ =>
+          false
+      }
+    }
+    def isScalacField: Boolean = {
+      val isField = sym.isInstanceOf[MethodSymbol] && !sym.isMethod && !sym.isParam
+      val isJavaDefined = sym.isJava
+      isField && !isJavaDefined
+    }
+    def isUselessField: Boolean = {
+      val peers = sym.parent.map(_.children.toList).getOrElse(Nil)
+      val getter = peers.find(m => m.isAccessor && m.name == sym.name.stripSuffix(" "))
+      sym.isScalacField && getter.nonEmpty
+    }
+    def isUsefulField: Boolean = {
+      sym.isScalacField && !sym.isUselessField
+    }
+    def isSyntheticCaseAccessor: Boolean = {
+      sym.isCaseAccessor && sym.name.contains("$")
+    }
+    def isRefinementClass: Boolean = {
+      sym.name == "<refinement>"
+    }
+    def isUseless: Boolean = {
+      sym == NoSymbol ||
+      sym.isAnonymousClass ||
+      sym.isSyntheticConstructor ||
+      sym.isModuleClass ||
+      sym.isLocalChild ||
+      sym.isExtensionMethod ||
+      sym.isSyntheticValueClassCompanion ||
+      sym.isUselessField ||
+      sym.isSyntheticCaseAccessor ||
+      sym.isRefinementClass
+    }
+    def isUseful: Boolean = !sym.isUseless
+  }
+
+  private var nextId = 0
+  private def freshSymbol(): String = {
```

Comment:
I'm a concerned about producing local symbols from metacp since those can get mixed with local symbols in a text document. For example, scalafix symtab is per-source which means after this PR, it's not clear how to implement `info("local1")` since `local1` may come from a hardlinked symbol from a dependency or from a local variable in the current text document.

I've wondered if we should adopt a similar scheme as JVM signatures where the first character indicates the type

* prefix global symbols with `G`
* prefix local symbols with `L`
* prefix hardlink symbols with `H`

Hardlinked symbols could additionally be suffixed by the enclosing owner symbol so that 

```diff
- this PR
+ proposal

advanced.Structural#s2(). => method s2: AnyRef { val def x: Int }
  AnyRef => scala.AnyRef#
-  x => local1
+  x => Hadvanced.Structural#s2().1
  Int => scala.Int#
```

This would mean hardlinked symbols could remain globally unique.